plugins/file-highlight/prism-file-highlight.js

```diff
@@ -1,89 +1,141 @@
+
+	var SELECTOR = 'pre[data-src]:not([' + STATUS_ATTR + '="' + STATUS_LOADED + '"])'
+		+ ':not([' + STATUS_ATTR + '="' + STATUS_LOADING + '"])';
+
+	var lang = /\blang(?:uage)?-([\w-]+)\b/i;
+
 	/**
-	 * @param {Element} [container=document]
+	 * Sets the Prism `language-xxxx` or `lang-xxxx` class to the given language.
+	 *
+	 * @param {HTMLElement} element
+	 * @param {string} language
+	 * @returns {void}
 	 */
-	self.Prism.fileHighlight = function(container) {
-		container = container || document;
-
-		var Extensions = {
-			'js': 'javascript',
-			'py': 'python',
-			'rb': 'ruby',
-			'ps1': 'powershell',
-			'psm1': 'powershell',
-			'sh': 'bash',
-			'bat': 'batch',
-			'h': 'c',
-			'tex': 'latex'
-		};
-
-		Array.prototype.slice.call(container.querySelectorAll('pre[data-src]')).forEach(function (pre) {
-			// ignore if already loaded
-			if (pre.hasAttribute('data-src-loaded')) {
-				return;
-			}
+	function setLanguageClass(element, language) {
+		var className = element.className;
+		className = className.replace(lang, ' ') + ' language-' + language;
+		element.className = className.replace(/\s+/g, ' ').trim();
+	}
 
-			// load current
-			var src = pre.getAttribute('data-src');
 
-			var language, parent = pre;
-			var lang = /\blang(?:uage)?-([\w-]+)\b/i;
-			while (parent && !lang.test(parent.className)) {
-				parent = parent.parentNode;
-			}
+	Prism.hooks.add('before-highlightall', function (env) {
+		env.selector += ', ' + SELECTOR;
+	});
 
-			if (parent) {
-				language = (pre.className.match(lang) || [, ''])[1];
-			}
+	Prism.hooks.add('before-sanity-check', function (env) {
+		var pre = /** @type {HTMLPreElement} */ (env.element);
+		if (pre.matches(SELECTOR)) {
+			env.code = ''; // fast-path the whole thing and go to complete
 
-			if (!language) {
-				var extension = (src.match(/\.(\w+)$/) || [, ''])[1];
-				language = Extensions[extension] || extension;
-			}
+			pre.setAttribute(STATUS_ATTR, STATUS_LOADING); // mark as loading
 
-			var code = document.createElement('code');
-			code.className = 'language-' + language;
+			// add code element with loading message
+			var code = pre.appendChild(document.createElement('CODE'));
+			code.textContent = LOADING_MESSAGE;
 
-			pre.textContent = '';
+			var src = pre.getAttribute('data-src');
 
-			code.textContent = 'Loading…';
+			var language = env.language;
+			if (language === 'none') {
+				// the language might be 'none' because there is no language set;
+				// in this case, we want to use the extension as the language
+				var extension = (/\.(\w+)$/.exec(src) || [, 'none'])[1];
+				language = EXTENSIONS[extension] || extension;
+			}
 
-			pre.appendChild(code);
+			// set language classes
+			setLanguageClass(code, language);
+			setLanguageClass(pre, language);
 
-			var xhr = new XMLHttpRequest();
+			// preload the language
+			var autoloader = Prism.plugins.autoloader;
+			if (autoloader) {
+				autoloader.loadLanguages(language);
+			}
```

Comment:
I keep coming back to this. Just to check my understanding: we don't need to coordinate the autoloader with this plugin because it doesn't matter which finishes first, both will just rehighlight after they're done. Is this right?

Reply:
No. Both work by 1) loading the file/data, and 2) highlighting it. 
This is just an optimization of step 2). If Autoloader was active, then it might have to download some languages in step 2). But we already know the language in step 1), so we preload them using `autoloader.loadLanguages(language)` (This doesn't do any highlighting). This means that both the language and the file/data can be downloaded in parallel. When Autoloader highlights the code in step 2), the language will already be loaded and highlighting can start right away.

Reply:
TBH, Autoloader might try to highlight the code block in step 1) but it should short-circuit because there isn't any code and if it doesn't then the "highlighted code" (whatever it is) will be overwritten in step 2).

So it's really no problem at all.

Reply:
Oh, I guess your understanding was correct. Just my understanding of your understanding wasn't...

But can humans truly ever understand each other? kappa